### ISSUE TITLE
$geoNear, $near, and $nearSphere are not allowed in this context 

### ISSUE THREAD

**Issue by jmurphy45** [jenssegers/laravel-mongodb#1819]
I am working on a project that gets an event near a certain gps cordinate and I am getting this error. I followed the documentation and running the latest 3.5 release.

```
 $events = Event::where('location', 'near', [
                '$geometry' => [
                    'type' => 'Point',
                    'coordinates' => [
                        $request->longitude,
                        $request->latitude,
                    ],
                ],
                '$maxDistance' => 30,
            ])->paginate(7);
```

`production.ERROR: $geoNear, $near, and $nearSphere are not allowed in this context {"userId":"5d335704802df000076bad97","exception":"[object] (MongoDB\\Driver\\Exception\\CommandException(code: 2): $geoNear, $near, and $nearSphere are not allowed in this context at /app/vendor/mongodb/mongodb/src/Operation/Aggregate.php:263)`

**Comment by Smolevich** [comment]
@jmurphy, You can join to [slack](https://join.slack.com/t/laravel-mongodb/shared_invite/enQtNzQyOTAzMzg0ODIwLWNjMzA0MjUwYzNiM2Q2NDdkYmE2NGJjZDM3NjgzODE2NTFlN2NlOWVjYjdiNjFjZTQ2ZjYwZDgzMGRiZGI4MGM) community and ask there

**Comment by Smolevich** [comment]
@jmurphy45, any updates?

**Comment by jmurphy45** [comment]
> @jmurphy45, any updates?

Not yet I haven't been able to find the issue.

**Comment by kanidjar** [comment]
Hi @Smolevich , I am getting this error too. Same result when using whereRaw.

MongoDB version: 4.2.1
PECL MongoDB version: 1.6.0
Package version: 3.6

```php

$this->where(
    'location', 
    'nearSphere', 
    [
        '$geometry' => [
            'type' => 'Point',
            'coordinates' => [
                $longitude,
                 $latitude,
             ],
        ],
        '$maxDistance' => $maxDistance,
    ]
);
```

![Screenshot 2019-10-27 at 14 06 53](https://user-images.githubusercontent.com/12607242/67635083-10762480-f8c3-11e9-9944-9f3599609e89.png)

Same query works when launched with MongoDB CLI:

![Screenshot 2019-10-27 at 14 08 09](https://user-images.githubusercontent.com/12607242/67635104-3ac7e200-f8c3-11e9-9cd7-25b15b412fe9.png)

Any idea?

**Comment by Smolevich** [comment]
@kanidjar, dump query with `DB::getQueryLog()`. Before dumping enable query log using `DB:: enableQueryLog`

**Comment by kanidjar** [comment]
@Smolevich I cannot dump the query since it throws an exception and the query is not visible in the exception.

However, i have been able to find out that the error occurs as soon as a count() or a paginate() is run with the $nearSphere condition

**Comment by Smolevich** [comment]
@kanidjar, you can use try-catch and in catch block dump query

**Comment by kanidjar** [comment]
@Smolevich 

```php
$queryBuilder = $this->getModelClass()::filter($request->all());

try {
    DB::connection('mongodb')->enableQueryLog();
    $count = $queryBuilder->count(); 
}
catch(\Exception $e) {
    dd(DB::connection('mongodb')->getQueryLog());
}
```

gives

```php
[]
```

The simplest way to reproduce the bug is to append a count() or a paginate() combined with a $nearSphere/$near condition

By the way, if you replace count() with get(), the query runs successfully.

**Comment by Smolevich** [comment]
@kanidjar, what does mean `$this->getModelClass()` ?

**Comment by kanidjar** [comment]
@Smolevich this code is part of an abstract controller in our project. Children controllers define their "Model class" through an abstract method called "getModelClass".

::filter() is a method exposed from tucker-eric/eloquentfilter package.

**Comment by Smolevich** [comment]
@kanidjar, i try to repeat your steps 
1. Create test model
```
use Jenssegers\Mongodb\Eloquent\Model;
use EloquentFilter\Filterable;

class Customer extends Model
{
    use Filterable;
    public $collection = 'customer';
    public $fillable = [
        'auth_id',
        'data',
    ];

    public function modelFilter()
    {
        return $this->provideFilter(\App\ModelFilters\CustomerFilter::class);
    }
}
```
CustomFilter:
```
namespace App\ModelFilters;

use EloquentFilter\ModelFilter;

class CustomerFilter extends ModelFilter
{
    /**
    * Related Models that have ModelFilters as well as the method on the ModelFilter
    * As [relationMethod => [input_key1, input_key2]].
    *
    * @var array
    */
    public $relations = [];
}

```
2. I create test code for getting count models with filter

```
        DB::connection('mongodb')->enableQueryLog();
        $filter = [
            'data.firstname' => 'John',
        ];
        $count = Customer::filter($filter)->count();
        dd(DB::connection('mongodb')->getQueryLog());
```
  Output:

```
array:1 [
  0 => array:3 [
    "query" => "customer.aggregate([{"$group":{"aggregate":{"$sum":1},"_id":null}}],{"typeMap":{"root":"array","document":"array"}})"
    "bindings" => []
    "time" => 847.97
  ]
]
```

**Comment by kanidjar** [comment]
@Smolevich I think you missed the information that the error occurs with $nearSphere condition!

Try to add the following filter to your custom filter:

```php
    public function closeTo($closeTo)
    {
        $data = explode(';', $closeTo);

        $latitude = $data[0];
        $longitude = $data[1];
        $maxDistance = $data[2];

        $this->where(
            'location', 
            'nearSphere', 
            [
                '$geometry' => [
                    'type' => 'Point',
                    'coordinates' => [
                        (float) $longitude,
                        (float) $latitude,
                    ],
                ],
                '$maxDistance' => (float) $maxDistance,
            ]
        );
    }
```

**Comment by Smolevich** [comment]
@kanidjar, please add more information 
1. one document from collection for test
2. Type and structure geo index

**Comment by kanidjar** [comment]
@Smolevich : here it is

```javascript
// collection: stations
{
	"_id" : ObjectId("5db58fdae0a92e715054b7a4"),
	"code" : "NF006062",
	"name" : "RELAIS PARIS CLICHY",
	"brand_id" : "TOT",
	"phone" : "0142291443",
	"address1" : "190 AVENUE DE CLICHY ",
	"address2" : null,
	"zip" : "75017",
	"city" : "PARIS",
	"country_id" : "FR",
	"currency" : "EUR",
	"location" : {
		"type" : "Point",
		"coordinates" : [
			2.315044,
			48.893808
		]
	},
	"open_status" : "2",
	"schedule" : {
		"week" : "06:00-22:00",
		"saturday" : "06:00-22:00",
		"sunday" : "07:00-22:00",
		"holidays" : "07:00-22:00"
	},
	"deleted_at" : null,
	"updated_at" : ISODate("2019-10-27T13:38:50.000+01:00"),
	"created_at" : ISODate("2019-10-27T13:38:50.000+01:00"),
	"item_codes" : [
		"aspirateur",
		"shop",
		"visa",
		"jubileo",
		"carwash",
		"carteclubtotal",
		"centrepneus",
		"superette",
		"Freewifi",
		"mastercard",
		"eurotrafic",
		"excellium98",
		"excelliumdiesel",
		"grnetherlands",
		"gasoil",
		"amex",
		"maintenancebays",
		"BusinessPro",
		"Speedy",
		"cartelavage",
		"CafeComptoirArome",
		"BouteillesGaz",
		"Lavage_MRN",
		"Aspirateur_MRN",
		"GRActysEurotrafic",
		"Lavage_Rouleau_MRN",
		"Banc",
		"SP95E10",
		"Gonflage",
		"LavageLOTUS",
		"PompePL",
		"stationtruck",
		"Justbip2",
		"MOL",
		"Bottlesadblue",
		"AGGLO",
		"GRcarteFR",
		"Monster",
		"mol"
	]
}
```


Indexes:

```javascript
db.stations.createIndex(
    {
        location: "2dsphere"
    },
    {
        name: "location_2dsphere",
        "2dsphereIndexVersion": 3
    }
)
```

**Comment by kanidjar** [comment]
@Smolevich , have you been able to reproduce the issue?

**Comment by parth7676** [comment]
@Smolevich Do you have pagination implemented ??

**Comment by Smolevich** [comment]
> @Smolevich , have you been able to reproduce the issue?
I can check and answer tomorrow

**Comment by Smolevich** [comment]
@kanidjar, i check this query and it works for me
php version: `7.3.10` 
mongodb php driver: `1.5.1`
MongoDB extension version: `1.6.0`

**Comment by kanidjar** [comment]
@Smolevich Could you please show me your Eloquent query?

**Comment by kanidjar** [comment]
@Smolevich great news, it looks like the bug has been fixed in version 3.6.1 :)

**Comment by StevePorter92** [comment]
3.6.1 fixed it for me also! Thank you
